Adding Robust Bayesian Meta-Analysis (beta)

FBartos opened the pull request:
The analysis fits an ensemble of meta-analytical models usings JAGS and then model-averages over them - so, the whole analysis can get quite time-consuming (talking about tens of minutes for datasets with 10-20 studies, or even hours with larger ones). Also, datasets with extreme datapoints might cause models to fail to converge (we use likelihood on the test-statistics instead of effect sizes which can be very hard to evaluate).

I included an already fitted model to easy the testing
https://drive.google.com/file/d/1n7nbDz7M9Ef3ryNscR4YtaPTOzS801sV/view?usp=sharing

And a small dataset
[MA_test.xlsx](https://github.com/jasp-stats/MetaAnalysis/files/4717249/MA_test.xlsx)

required files
https://github.com/FBartos/jasp-required-files/tree/Windows-MM

Kucharssim commented:
I am on a Mac, is there `jasp-required-files` that can run this or do you only have a version for Windows? 

I can alternatively check it out on my old Windows computer, but it will be annoying for me, and it's always good to test on different platforms anyway.

FBartos commented:
Hi Simon, I will add the required packages for installation. I just for reviews for Mixed Models and there are some changes that will be required for this module as well. I will make them ASAP, but could you postpone your review until then, please? I will need a day or two.

Kucharssim commented:
Sure, take your time!

FBartos commented:
ok, I added the required changes and it's ready for review.

The required packages are: `"runjags", "extraDistr", "DPQ"` and 
`devtools::install_github("https://github.com/FBartos/RoBMA")`

vandenman commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2419 @@
+      name = "upperRange",
+      title = gettext("Upper"),
+      type = "number",
+      overtitle = paste0(gettext("<em>p</em>-values interval"), "\u002A")
```
`gettextf`

vandenman commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2419 @@
+    return(jasp_table)
+  }
+.RoBMA_coef_names           <- function(name, add_info) {
+  if (substr(name, 1, 2) == "mu")
```
why is this approach with `substr` necessary? It seems to me that this is incredibly sensitive to the renaming of e.g., model elements. Can't we derive this from user input?

FBartos replied:
That would be too complicated since the function is used also for the individual models - I would have to reconstruct which model does and doesn't have the specific parameters. The function is not used as widely as it was designed to previously, so I removed the `substr()` part at least.

vandenman commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2419 @@
+    )
+  }
+}
+.RoBMA_options2priors_clean <- function(x) {
```
What's happening in this function? What's the magic number 666? I have no idea what's going on here. The while loop also looks quite dangerous.

FBartos replied:
Honestly? I have no idea. Probably an easter egg :D 
(removed)

FBartos replied:
I mean the loop with 666.
The rest is just to clean input from the users if they happen to specify the vector of cutoffs / arguments for the Dirichlet distribution in R-like format.

` 
x <- trimws(x, which = "both")  

x <- trimws(x, which = "both", whitespace = "c")  

x <- trimws(x, which = "both", whitespace = "\\(")  

x <- trimws(x, which = "both", whitespace = "\\)")  

x <- trimws(x, which = "both", whitespace = ",")  
`

vandenman commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2419 @@
+        gettext("correspond to effect size"),
+        "\u03C1."
+      ))
+    if (add_info$effect_size == "d")
```
if `if (add_info$effect_size == "d")` and `if (add_info$effect_size == "r")` a couple lines above are mutually exclusive then this should be an `else if`.

FBartos replied:
yes, you are right. changed

vandenman commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2419 @@
+  
+  
+  # add priors from options
+  if (length(options$priors_mu) != 0) {
+    for (i in 1:length(options$priors_mu)) {
+      priors_mu <- c(priors_mu,
+                     list(.RoBMA_options2priors(options$priors_mu[[i]])))
+    }
+  }
+  
+  if (length(options$priors_tau) != 0) {
+    for (i in 1:length(options$priors_tau)) {
+      priors_tau <- c(priors_tau,
+                      list(.RoBMA_options2priors(options$priors_tau[[i]])))
+    }
+  }
```
something like this
```r
object <- list()
prior_elements <- c("priors_mu", "priors_omega", ...)
prior_names <- c("mu", "omega", ...)
for (i in seq_along(prior_elements)) {
  tmp <- NULL
  for (elem in options[[prior_elements[i]]]) {
     tmp <- c(tmp, list(.RoBMA_options2priors(elem)))
  }
  object[[prior_names[i]]] <- tmp
}
```
is equivalent to the 6 for loops right? if you use `seq_along` or `elem in ...` then the empty lists are automatically skipped.

FBartos replied:
That's very clever! Thanks, I changed it accordingly.

JorisGoosen commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2419 @@
+    }
+    
+    if (options$measures == "fitted") {
+      fit <- readRDS(file = options$fitted_path)
```
Should we make an issue for it? 
Then once implemented we can also replace this code.

FBartos commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2419 @@
+    
+    if (options$measures == "fitted") {
+      fit <- readRDS(file = options$fitted_path)
+      if (class(fit) != "RoBMA")
```
ok, changed as suggested

FBartos commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2419 @@
+    
+    # add footnote
+    if (add_info$effect_size == "r")
+      jasp_table$addFootnote(paste0(
```
addressed, and all of the following `gettextf` as well

FBartos commented:
Thanks for the comments and suggestions Don. I will address the error & disappearing table in the following commits (it's not disappearing on my PC, which is kinda weird because it should according to the code, and I have the same problem with the Main summary table when changing input.).

![image](https://user-images.githubusercontent.com/38475991/83806830-5f363900-a6b2-11ea-9725-dc17caf52883.png)

**fitting times**
The ensemble is unfortunately complex, therefore the fitting time is high. However, I think that researchers should be able to wait a few hours to get the results if they spent a few months recording the studies and etc. 

For the testing, the first comment contains a dataset with 3 studies that can be fitted in a reasonable time. Four minutes and ten seconds with the default settings, or one minute and forty seconds when decreasing the number of chains to 2 and burning and iteration to 4000 (under 'Advanced options' tab). The time can be further greatly reduced by removing one of the priors for selection bias under the 'Priors' tab. Another option is to load the already fitted model (link also in the first comment).

FBartos commented:
OK, so the last commits should address the rest of the problems.
1) The model preview table should be visible during the fitting process (It worked for me before, so I hope that this change fixes it for everyone).
2) The bug with the debug dataset was due to the fact that the combination of Cohen's d and its standard error would mean that some sample sizes were lower than 1. I changed the error message to be more informative (with the new package version, not in the required files yet). Now, I also try adding the error messages on the model preview table instead of directly crashing the analysis.

I'm not sure whether it makes more sense to test the analysis once it's merged into JASP as discussed on the Thursday meeting, which should happen on Monday).

FBartos commented:
Hi @Kucharssim and @vandenman,
I talked with Joris and I misunderstood how we wanted to proceed with the integration. The plan is to first finish the reviews here and then merge it. Could you please take a look at the module again? I think I addressed all of the issues raised before. I updated the RoBMA package, so you need to either reinstall it `devtools::install_github("https://github.com/FBartos/RoBMA")`.

small recap:
there is an already fitted model that can be loaded for testing:
https://drive.google.com/file/d/1n7nbDz7M9Ef3ryNscR4YtaPTOzS801sV/view?usp=sharing
or a small dataset with only three studies:
[MA_test.txt](https://github.com/jasp-stats/MetaAnalysis/files/4746436/MA_test.txt)
(it takes 4'10'' with the default settings, or 1'40'' when decreasing the number of chains to 2 and burning and iteration to 4000 (under 'Advanced options' tab). The time can be further greatly reduced by removing one of the priors for selection bias under the 'Priors' tab)

Kucharssim commented:
I am not sure how much time I have today, but I will try to review asap - hopefully tomorrow the latest.

Kucharssim commented:
Analysis crashed after loading the Anderson_exp file and selecting "Conditional estimates". JASP file attached
[roBMA_bug.jasp.zip](https://github.com/jasp-stats/MetaAnalysis/files/4757487/roBMA_bug.jasp.zip)

Kucharssim commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2299 @@
+  
+  return()
+}
+.RoBMA_plots                <-
```
This function is a nesting hell, any chance of refactoring it into separate smaller functions?

Kucharssim replied:
Never mind, that seems to be an overall style of the plotting functions... Would be nice to thing about simplifying the code, but I understand it's not a priority for now

FBartos commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2299 @@
+          terms    = .RoBMA_coef_names(rownames(results_table)[i], add_info),
+          mean     = results_table$Mean[i],
+          median   = results_table$Median[i],
+          lowerCI  = results_table[i, 3],
```
It's not possible to do it like that. The tables are generated with the quantiles as column names because the summary function allows specifying quantiles for the summary (not only lower and upper CI, and it doesn't make sense to have different handling for CI and quantities. I think that rjags and rstan function return tables in similar matter.)

Kucharssim replied:
But you could construct the names if you know the specified credible %, the same way you do it internally in the package, right? I am not trying to be picky, just thinking about protecting against the future bug when you decide to add a Mode, SE, SD, or what have you to the table output, shifting the indexes of the intervals, having you to go through this code again and fixing the indexes.

FBartos commented on inst/qml/RobustBayesianMetaAnalysis.qml:
```diff
@@ -0,0 +1,2001 @@
+						}
+						FormulaField
+						{
+							label:				"α "
```
because the point and Cauchy prior distribution have x[0] as a first parameter, this makes the labels and boxes seem more aligned if two different distributions are selected

FBartos commented on inst/qml/RobustBayesianMetaAnalysis.qml:
```diff
@@ -0,0 +1,2001 @@
+			name:		"advanced_mu_transform"
+			values:
+			[
+				{ label: "Cohen's d",			value: "cohens_d"},
```
yes, thank you. I seem to have missed some of those

FBartos commented on inst/help/robustbayesianmetaanalysis.md:
```diff
@@ -0,0 +1,211 @@
+  - Weight function: Combine the weights into a weight function and display that instead.
+
+#### Type
+- Model-averaged: Pooled estimates will contain model-averaged estimates across all models.
```
no, we use model-averaging with the hyphen consistently in the paper and package output

Kucharssim replied:
But then that's inconsistent with other files in JASP

Kucharssim replied:
Just to be clear: I don't mind either way, but I prefer consistency between JASP modules above consistency between a paper and JASP module

FBartos commented on inst/help/robustbayesianmetaanalysis.md:
```diff
@@ -0,0 +1,211 @@
+### Input
+---
+#### Input type
+- Cohen's d / t-statistics & (N / SE): Specifying input using either Cohen's d measured effect sizes and the sample size or standard errors, or reported t-statistics and sample sizes.
```
true, this sounds better

FBartos commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2299 @@
+    if (any(rownames(results_table) == "tau")) {
+      if (add_info$effect_size == "r")
+        jasp_table$addFootnote(
+          gettextf("%s is on %s scale.", "\u03C4",ifelse(
```
they are exclusive but I agree that if else is more explicit

vandenman commented on R/robustbayesianmetaanalysis.R:
```diff
@@ -0,0 +1,2299 @@
+          upperCI  = results_table[i, 4]
+        )
+        if (individual) {
+          temp_row$error <- results_table$`MCMC error`[i]
```
@Kucharssim `[[` is indeed preffered over `$` although many people don't do that.
